File: 백준/Platinum/2887. 행성 터널/행성 터널.py
import sys
I=sys.stdin.readline
def f(x):
 if p[x]!=x:p[x]=f(p[x])
 return p[x]
def u(x,y):
 x,y=f(x),f(y)
 if x!=y:p[y]=x;return 1
 return 0
N=int(I())
P=[list(map(int,I().split()))+[i] for i in range(N)]
E=[]
for j in range(3):
 P.sort(key=lambda x:x[j])
 for i in range(N-1):E.append((abs(P[i][j]-P[i+1][j]),P[i][3],P[i+1][3]))
E.sort()
p=list(range(N))
c=n=0
for cost,i,j in E:
 if u(i,j):
  c+=cost;n+=1
  if n==N-1:break
print(c)

File: 백준/Platinum/2887. 행성 터널/행성 터널.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <tuple>

using namespace std;

struct Planet {
    int x, y, z;
    int index;
};

vector<int> parent;

int find(int u) {
    if (parent[u] != u)
        parent[u] = find(parent[u]);
    return parent[u];
}

void unite(int u, int v) {
    u = find(u);
    v = find(v);
    if (u != v)
        parent[v] = u;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int N;
    cin >> N;

    vector<Planet> planets(N);
    for (int i = 0; i < N; ++i) {
        cin >> planets[i].x >> planets[i].y >> planets[i].z;
        planets[i].index = i;
    }

    vector<tuple<int, int, int>> edges;

    sort(planets.begin(), planets.end(), [](const Planet& a, const Planet& b) {
        return a.x < b.x;
    });
    for (int i = 1; i < N; ++i) {
        int cost = abs(planets[i].x - planets[i-1].x);
        edges.emplace_back(cost, planets[i-1].index, planets[i].index);
    }

    sort(planets.begin(), planets.end(), [](const Planet& a, const Planet& b) {
        return a.y < b.y;
    });
    for (int i = 1; i < N; ++i) {
        int cost = abs(planets[i].y - planets[i-1].y);
        edges.emplace_back(cost, planets[i-1].index, planets[i].index);
    }

    sort(planets.begin(), planets.end(), [](const Planet& a, const Planet& b) {
        return a.z < b.z;
    });
    for (int i = 1; i < N; ++i) {
        int cost = abs(planets[i].z - planets[i-1].z);
        edges.emplace_back(cost, planets[i-1].index, planets[i].index);
    }

    sort(edges.begin(), edges.end());

    parent.resize(N);
    for (int i = 0; i < N; ++i)
        parent[i] = i;

    int total = 0;
    for (auto& edge : edges) { // C++11 compatible iteration
        int cost = get<0>(edge);
        int u = get<1>(edge);
        int v = get<2>(edge);
        if (find(u) != find(v)) {
            unite(u, v);
            total += cost;
        }
    }

    cout << total;
    return 0;
}